### ISSUE TITLE
Некоторые игры не работают .

### ISSUE THREAD

**Issue by igor4555** [NyashniyVladya/Translator3000#3]
При запуске некоторых игр с программой translator3000 игры вылетают с такой ошибкой .
Можете проверить на игре Elysion и Rainy Day . 
А так программа классная .

I'm sorry, but an uncaught exception occurred.

While running game code:
  File "game/Translator3000Data/scripts/imports.rpy", line 17, in <module>
TypeError: 'dict_keys' object does not support indexing

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "renpy/bootstrap.py", line 326, in bootstrap
    renpy.main.main()
  File "renpy/main.py", line 511, in main
    renpy.game.context().run(node)
  File "Translator3000Scripts/imports.rpyc", line 2, in script
  File "Translator3000Scripts/imports.rpyc", line 2, in script
  File "renpy/ast.py", line 908, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 2104, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "game/Translator3000Data/scripts/imports.rpy", line 17, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "_translator3000/__init__.py", line 6, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "requests/__init__.py", line 52, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "requests/packages/__init__.py", line 27, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "requests/packages/urllib3/__init__.py", line 8, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "requests/packages/urllib3/connectionpool.py", line 35, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "requests/packages/urllib3/connection.py", line 18, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "tlslite/__init__.py", line 27, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "tlslite/api.py", line 11, in <module>
  File "renpy/loader.py", line 953, in load_module
    exec(code, mod.__dict__)
  File "tlslite/tlsconnection.py", line 23, in <module>
TypeError: 'dict_keys' object does not support indexing

**Comment by NyashniyVladya** [comment]
"Rainy Day" выпущена на бета-версии Ren'Py 7.4.0, которая в данный момент [на активной стадии багофиксов](https://lemmasoft.renai.us/forums/viewtopic.php?t=60461).
Зачем, несмотря на предупреждение Пайтома (разработчик движка) не публиковать игры, собранные на 7.4.0,  разработчик игры это сделал - непонятно.

Первую игру я не проверял, но, если ошибка та же самая, - скорее всего, она тоже на 7.4.0.

Можете посмотреть версию Ren'Py в файле **_renpy/\_\_init\_\_.py_** в директории игры.
Открываете файл и листаете до объявления переменной **_version_tuple_**.

Поддержку 7.4.0 я, разумеется, планирую, но только после того, как она выйдет из бета-теста.

**Comment by igor4555** [comment]
Да ошибка такая же , Спасибо , буду ждать обновленной версии переводчика .

**Comment by MrAbsolutely** [comment]
Некоторые игры и на версии 7.3.5 не захватываются. Because I Love Her как пример.
![image](https://user-images.githubusercontent.com/76562337/103370929-f2700980-4aef-11eb-982d-7db5d3da0889.png)

**Comment by NyashniyVladya** [comment]
@MrAbsolutely 
Если не затруднит, поставьте  в настройках значение **__debug_mode_** на **_true_** и скиньте текст ошибки, которая вылетит после этого.

**Comment by MrAbsolutely** [comment]
[lastLog.log](https://github.com/NyashniyVladya/Translator3000/files/5775731/lastLog.log)
И где можно почитать список изменений программы? А то новые версии выходят и не понятно что в них.

**Comment by NyashniyVladya** [comment]
> И где можно почитать список изменений программы? А то новые версии выходят и не понятно что в них.

На [странице коммитов](https://github.com/NyashniyVladya/Translator3000/commits).
Но в последние версий 10 я ничего глобального не добавлял. Исправления небольшие и рефакторинг старого кода Последнее обновление было посвящено графическому интерфейсу, но сам функционал тот же что и был.

> [lastLog.log](https://github.com/NyashniyVladya/Translator3000/files/5775731/lastLog.log)

Файл **_traceback.txt_** ещё можно поглядеть?

**Comment by MrAbsolutely** [comment]
А где traceback.txt находится? Не вижу его в папке с игрой.

**Comment by NyashniyVladya** [comment]
> А где traceback.txt находится? Он должен быть с ярлыком запуска игры, но т.к. ошибки не было его нету. Не вижу его в папке с игрой.

Тогда это вообще странно, если ошибок не было... При включённом режиме дебага переводчик сигнализирует обо всех неполадках. Настройки точно в порядке? Направление перевода и исходный язык не равны одному значению, например?
Попозже скачаю эту игру, попробую воспроизвести проблему.

**Comment by MrAbsolutely** [comment]
Я сейчас кое что проверю и скажу. Похоже я догадываюсь в чём проблема.

**Comment by MrAbsolutely** [comment]
Нет, не получается. Удалил все сохранения, так же в C:\Users\02132\AppData\Roaming\RenPy удалил папку. Ничего. У меня один раз было такое, что играл сначала в русскую версию, там язык на рус поставил. Затем продолжил англ. версию. Но перевод не осуществлялся. Приходилось файлы удалять.
Так же я заметил, что когда начал пользоваться этой программкой, то игры RenPy очень часто не запускаются с первого раза. Даже если программа не установлена в данную игру.

**Comment by NyashniyVladya** [comment]
> Так же я заметил, что когда начал пользоваться этой программкой, то игры RenPy очень часто не запускаются с первого раза. Даже если программа не установлена в данную игру.

А это уже совсем странно. Без **_.rpa_** архива в папке она не оказывает никакого влияния на игру и не прописывает себя ни в реестры, ни в постоянные файлы ренпая.
P.S.: Не прописыва**_ла_**, если быть точнее. Версия с сегодняшнего обновления таки прописывает себя в multi persistent для переноса настроек между играми, но это, опять же, никак не может влиять на игру, если **_.rpa_** нет в директории.

**Comment by MrAbsolutely** [comment]
Нашел еще одну критическую ошибку: перестали работать сохранения во всех играх, где установлен переводчик. Т.е. нельзя сохраниться
[traceback.txt](https://github.com/NyashniyVladya/Translator3000/files/5776339/traceback.txt)

**Comment by NyashniyVladya** [comment]
Да, мне уже известно об этом. Не обнаружил на этапе отладки. В ближайшее время починю.

**Comment by NyashniyVladya** [comment]
Проблема с сохранением исправлена в последнем обновлении.

**Comment by Svyatoslav2** [comment]
Хотелось бы узнать,старые версии RenPy ведь тоже поддерживаются переводчиком,а то у меня выдает такой лог прога.
![Снимок](https://user-images.githubusercontent.com/51791931/104814845-fb681700-5821-11eb-9110-8dbc96939aae.JPG)

**Comment by NyashniyVladya** [comment]
> Хотелось бы узнать,старые версии RenPy ведь тоже поддерживаются переводчиком...

Поддерживаются версии, начиная с 6.99.12.4 (включительно) и новее (не считая 7.4.0 - её поддержку я реализую, как только выйдет стабильная версия, ибо текущая слишком сырая и весьма багованная).
Более старые версии ренпая плохо дружат с внедрением стороннего python кода. 6.99.14 - крайний предел обратной совместимости, которого я смог добиться.

**Comment by Svyatoslav2** [comment]
Очень жаль,но что же поделать,придется видимо чисто с английским языком играть(

**Comment by Divers102** [comment]
[traceback.txt](https://github.com/NyashniyVladya/Translator3000/files/5847284/traceback.txt)

**Comment by Divers102** [comment]
Игра не запускается, а игра называется Ecchi Sketch v1.0 (All Platforms)

**Comment by NyashniyVladya** [comment]
> [traceback.txt](https://github.com/NyashniyVladya/Translator3000/files/5847284/traceback.txt)

Ровно та же ситуация, что и с постом выше. Версия игры 6.99.10.1227.

**Comment by Divers102** [comment]
Понятно

**Comment by RosFighter** [comment]
Когда примерно планируется поддержка RenPy 7.4? Новые версии многих игру выходят уже на этом движке

**Comment by NyashniyVladya** [comment]
> Когда примерно планируется поддержка RenPy 7.4? Новые версии многих игру выходят уже на этом движке

Да, у меня уже есть это в планах. В ближайшее время займусь.

**Comment by HEB3OP** [comment]
Доброго дня, спасибо за такой отличный переводчик! Столкнулся со странной проблемой, при первом запуске игры выдаёт такую ошибку:
`I'm sorry, but an uncaught exception occurred.

Before loading the script.
Exception: Couldn't find file 'Translator3000.rpi'.

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "C:\HEB3OP\Torrent\Patronus-Trunk-winnozip\renpy\bootstrap.py", line 314, in bootstrap
    renpy.main.main()
  File "C:\HEB3OP\Torrent\Patronus-Trunk-winnozip\renpy\main.py", line 308, in main
    renpy.loader.index_archives()
  File "C:\HEB3OP\Torrent\Patronus-Trunk-winnozip\renpy\loader.py", line 164, in index_archives
    fn = transfn(prefix + ".rpi")
  File "C:\HEB3OP\Torrent\Patronus-Trunk-winnozip\renpy\loader.py", line 649, in transfn
    raise Exception("Couldn't find file '%s'." % name)
Exception: Couldn't find file 'Translator3000.rpi'.

Windows-8-6.2.9200
Ren'Py 7.3.2.320
 
Tue Feb 16 10:42:23 2021
`

**Comment by NyashniyVladya** [comment]
> `I'm sorry, but an uncaught exception occurred.
> 
> Before loading the script.
> Exception: Couldn't find file 'Translator3000.rpi'.

Это действительно очень странная ошибка...
**_.rpi_** не используется со времён пятой версии ренпая, а здесь, как я вижу, 7.3.2.320... Я попробую воспроизвести проблему и разобраться.

**Comment by MrAbsolutely** [comment]
Версия 2.5.10 работает не во всех играх в RenPy в версиях выше 7,4,0. В игре Max slive выскакивает такая ошибка на старте.
[traceback.txt](https://github.com/NyashniyVladya/Translator3000/files/6046970/traceback.txt)

**Comment by NyashniyVladya** [comment]
> [traceback.txt](https://github.com/NyashniyVladya/Translator3000/files/6046970/traceback.txt)

Вы уверены, что версия Ren'Py выше 7.4.0? Такая ошибка может возникать, если Вы скопируете файл для новых версий в игру со старой (<=7.3.5) версией.

Для старых версий идёт отдельный файл и пояснения в инструкции (решение временное, позже сделаю универсальный).

**Comment by MrAbsolutely** [comment]
Да, уверен. Скачал универсальную версию и всё равно такая же ошибка.
![image](https://user-images.githubusercontent.com/76562337/109301460-d2dc2f80-7859-11eb-95cf-18f23302a02f.png)
![image](https://user-images.githubusercontent.com/76562337/109301557-f4d5b200-7859-11eb-87a0-33a01edeee13.png)

**Comment by MrAbsolutely** [comment]
Еще одна игра ошибку выдает такую же. 
https://f95zone.to/threads/vae-victis-conquer-ravish-breed-v0-1-19-doorknob22.74639/

**Comment by viewerigo** [comment]
Огромное вам спасибо за прекрасную программу!

Не хватает только перевода от Яндекса - он сейчас выдаёт перевод значительно лучше качеством, чем Гугл. Но однозначно лучше гуглоперевод, чем полное его отсутствие!

А можете ответить на такой глупый вопрос: где сохраняются результаты перевода?

**Comment by Svyatoslav2** [comment]
Вопрос,с недавнего времени почему то,игры которые должны переводится,просто стирают какой либо текст,то есть окна с текстом пустые,как это исправить?

**Comment by HEB3OP** [comment]
> Вопрос,с недавнего времени почему то,игры которые должны переводится,просто стирают какой либо текст,то есть окна с текстом пустые,как это исправить?

Читай инструкцию лучше... Это означает, что встроенные в игру шрифты не поддерживают кириллицу. Нужно свои шрифты ставить. Качаешь любые, кидаешь в ту же папку, что и файл переводчика, в настройках, в параметре "font" прописываешь название файла шрифта. Так же подбери размер шрифта в параметре "size".

**Comment by Svyatoslav2** [comment]
> > Вопрос,с недавнего времени почему то,игры которые должны переводится,просто стирают какой либо текст,то есть окна с текстом пустые,как это исправить?
> 
> Читай инструкцию лучше... Это означает, что встроенные в игру шрифты не поддерживают кириллицу. Нужно свои шрифты ставить. Качаешь любые, кидаешь в ту же папку, что и файл переводчика, в настройках, в параметре "font" прописываешь название файла шрифта. Так же подбери размер шрифта в параметре "size".

Я все так и делаю.всегда,но не работает

**Comment by Svyatoslav2** [comment]
Что же я понял,игры в которых в папке data лежат файлы,типа:rpyc не переводятся(

**Comment by tigerjkeee** [comment]
While running game code:
  File "game/script.rpy", line 18, in script
    init 1 python in _compile_build_init:
  File "game/script.rpy", line 23, in <module>
    raise Exception(__("Версии сборки и проекта не равны."))
Exception: Версии сборки и проекта не равны.

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "game/script.rpy", line 18, in script
    init 1 python in _compile_build_init:
  File "E:\Friendship with Benefits\renpy\ast.py", line 912, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "E:\Friendship with Benefits\renpy\python.py", line 2004, in py_exec_bytecode
    exec bytecode in globals, locals
  File "game/script.rpy", line 23, in <module>
    raise Exception(__("Версии сборки и проекта не равны."))
Exception: Версии сборки и проекта не равны.

Windows-8-6.2.9200
Ren'Py 7.3.2.320
Friendship with Benefits 1.20

Версия 7,3,2 а пишет что не совпадает, что может быть не так?

**Comment by NyashniyVladya** [comment]
@tigerjkeee 
Эммм... Это очень странная ситуация. Файла **_script.rpy_** в конечной билде нет. Он присутствует в репозитории для компиляции **_.rpa_**. Что именно Вы хотите сделать? Пересобрать проект?

**Comment by tigerjkeee** [comment]
скачал версию 2.6.0 там файл script.rpy есть, ругается на него, но если его удалить все равно тоже самое

**Comment by MrAbsolutely** [comment]
Игра Esteria [Ch.3] [Royal Games], перевод в игре полностью не работает. Когда он работает, то в главном меню во всех играх есть небольшое подлагивание. Здесь такого нету. 
В игре Max's Life [Ch. 4 v0.36] [Kuggazer] при запуске ошибка выскакивает.

**Comment by Feans** [comment]
Некоторые старые игры всё же можно перевести. Для этого нужно заменить в старой игре папку "renpy" (и возможно "lib") аналогичными, взятыми из какой-нибудь свежей игры на renpy. Таким образом мы получим "новый" движок, с которым translator3000 будет работать.

Это работает не со всеми играми. В некоторых играх возможны глюки типа некорректного позиционирования элементов на экране, или просто не запустится, но если повезёт, всё будет работать без особых последствий.

**Comment by MrAbsolutely** [comment]
> Некоторые старые игры всё же можно перевести. Для этого нужно заменить в старой игре папки "renpy" (и возможно "lib") аналогичными, взятыми из какой-нибудь свежей игры на renpy. Таким образом мыим "новый" движок, с которым получим переводчик3000 будет работать.
> 
> Это работает не со всеми играми. В некоторых играх возможны глюки типа некорректного позиционирования элементов на экране, или просто не запустится, но если повезёт, всё будет работать без особых последствий.

Работает этот способ. Max's Life перестал при запуске выдавать ошибку.

**Comment by Feans** [comment]
Ну в общем я пробовал только на двух-трёх играх пятилетней давности. Одна работает без проблем, а у второй текстура бэкграунда нижней плашки, в которой выводится весь текст растянулась по высоте на половину экрана. Третью не помню, запустилась она или нет.

**Comment by fuhgb** [comment]
Здравствуйте, я установил последнюю версию переводчика на Ren'Py 7.4.0.1167.

Но у меня почему-то появляется такой текст.

I'm sorry, but an uncaught exception occurred.

While running game code:
  File "game/Translator3000Data/scripts/general.rpy", line 608, in <module>
  File "game/Translator3000Data/scripts/general.rpy", line 101, in turn_on
  File "game/Translator3000Data/scripts/general.rpy", line 58, in __init__
  File "game/Translator3000Data/scripts/imports.rpy", line 112, in save
WindowsError: [Error 123] �R�y�~���p�{���y���u���{�p�� �����y�q�{�p �r �y�}�u�~�y ���p�z�|�p, �y�}�u�~�y ���p���{�y �y�|�y �}�u���{�u �����}�p: u'C:\\Users\\\udc84E\udc84|\udc84u\udc84~\udc84p'

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "renpy/bootstrap.py", line 326, in bootstrap
    renpy.main.main()
  File "renpy/main.py", line 515, in main
    renpy.game.context().run(node)
  File "Translator3000Scripts/general.rpyc", line 2, in script
  File "Translator3000Scripts/general.rpyc", line 2, in script
  File "renpy/ast.py", line 913, in execute
    renpy.python.py_exec_bytecode(self.code.bytecode, self.hide, store=self.store)
  File "renpy/python.py", line 2111, in py_exec_bytecode
    exec(bytecode, globals, locals)
  File "game/Translator3000Data/scripts/general.rpy", line 608, in <module>
  File "game/Translator3000Data/scripts/general.rpy", line 101, in turn_on
  File "game/Translator3000Data/scripts/general.rpy", line 58, in __init__
  File "game/Translator3000Data/scripts/imports.rpy", line 112, in save
  File "_translator3000/utils.py", line 53, in save
  File "_translator3000/utils.py", line 70, in _save
  File "_translator3000/utils.py", line 66, in _create_dir_for_file
  File "/home/tom/ab/renpy-build/tmp/install.android-x86_64/lib/python2.7/os.py", line 150, in makedirs
  File "/home/tom/ab/renpy-build/tmp/install.android-x86_64/lib/python2.7/os.py", line 150, in makedirs
  File "/home/tom/ab/renpy-build/tmp/install.android-x86_64/lib/python2.7/os.py", line 157, in makedirs
WindowsError: [Error 123] �R�y�~���p�{���y���u���{�p�� �����y�q�{�p �r �y�}�u�~�y ���p�z�|�p, �y�}�u�~�y ���p���{�y �y�|�y �}�u���{�u �����}�p: u'C:\\Users\\\udc84E\udc84|\udc84u\udc84~\udc84p'

и появляется только папка _translator3000Data

![image](https://user-images.githubusercontent.com/86412957/123260726-91ede200-d4fe-11eb-9554-c5382a40ce79.png)
 
Не могли бы вы мне подсказать что делать в этом случае.

**Comment by Divers102** [comment]
Здравствуйте, у меня игра запускается, но после нажатие начать играть, появляется ошибка.
Вот 
[traceback.txt](https://github.com/NyashniyVladya/Translator3000/files/7855183/traceback.txt)
и 
[lastLog.log](https://github.com/NyashniyVladya/Translator3000/files/7855185/lastLog.log)
и лог 
[log.txt](https://github.com/NyashniyVladya/Translator3000/files/7855187/log.txt)
